File: benchmarks/go/04_matrix_multiply.go
package main

import (
	"fmt"
	"time"
)

func main() {
	n := 200
	iterations := 3

	matA := make([][]int, n)
	matB := make([][]int, n)
	for i := 0; i < n; i++ {
		matA[i] = make([]int, n)
		matB[i] = make([]int, n)
		for j := 0; j < n; j++ {
			matA[i][j] = (i*n + j) % 100
			matB[i][j] = (i*n + j + 50) % 100
		}
	}

	times := make([]float64, 0, iterations)
	var checksum int

	for iter := 0; iter < iterations; iter++ {
		start := time.Now()

		result := make([][]int, n)
		for i := 0; i < n; i++ {
			result[i] = make([]int, n)
			for j := 0; j < n; j++ {
				val := 0
				for k := 0; k < n; k++ {
					val += matA[i][k] * matB[k][j]
				}
				result[i][j] = val
			}
		}

		elapsed := time.Since(start).Seconds() * 1000
		checksum = result[0][0]
		times = append(times, elapsed)
	}

	best := times[0]
	sum := 0.0
	for _, t := range times {
		if t < best {
			best = t
		}
		sum += t
	}
	avg := sum / float64(len(times))

	fmt.Println("BENCHMARK: matrix_multiply")
	fmt.Printf("size=%dx%d, iterations=%d\n", n, n, iterations)
	fmt.Printf("checksum=%d\n", checksum)
	fmt.Printf("best=%.6fms\n", best)
	fmt.Printf("avg=%.6fms\n", avg)
}

File: benchmarks/python/04_matrix_multiply.py
import time

n = 200
iterations = 3

mat_a = [[(i * n + j) % 100 for j in range(n)] for i in range(n)]
mat_b = [[(i * n + j + 50) % 100 for j in range(n)] for i in range(n)]

times = []
checksum = 0

for _ in range(iterations):
    start = time.perf_counter()

    result = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            val = 0
            for k in range(n):
                val += mat_a[i][k] * mat_b[k][j]
            result[i][j] = val

    elapsed = (time.perf_counter() - start) * 1000
    checksum = result[0][0]
    times.append(elapsed)

best = min(times)
avg = sum(times) / len(times)
print("BENCHMARK: matrix_multiply")
print(f"size={n}x{n}, iterations={iterations}")
print(f"checksum={checksum}")
print(f"best={best}ms")
print(f"avg={avg}ms")
